native port: Mirror the zephyr thread name onto native threads

rugeGerritsen opened the pull request:
When debugging native applications it is useful to know which thread is being debugged. Currently, all thread names are being displayed as "zephyr", making this hard.
The changes in this PR changes this to display the zephyr thread names, given that `CONFIG_THREAD_NAME` is enabled.

To avoid adding posix-specific code in the kernel, this is handled by adding a new optional arch API to the kernel interface.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/52223.

rugeGerritsen committed:
kernel: Define optional arch_thread_name_set()

The intention of this API is to allow setting the posix thread
name equal to the zephyr thread name.
By defining it as an arch interface, the implementation becomes
generic.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
arch posix: Implement arch_thread_name_set()

This will update the posix thread names to match
the zephyr thread names.

This will simplify debugging as the debugger will
recognize the thread names.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

aescolar commented on arch/posix/core/posix_core.c:
```diff
@@ -482,6 +482,18 @@ int posix_arch_get_unique_thread_id(int thread_idx)
+{
+	int err;
+
+	err = pthread_setname_np(&threads_table[thread_idx].thread, str);
```
```suggestion
	err = pthread_setname_np(threads_table[thread_idx].thread, str);
```

aescolar reviewed (changes requested):
Thanks @rugeGerritsen , see comments

rugeGerritsen commented on kernel/thread.c:
```diff
@@ -608,6 +619,7 @@ char *z_setup_new_thread(struct k_thread *new_thread,
 			CONFIG_THREAD_MAX_NAME_LEN - 1);
 		/* Ensure NULL termination, truncate if longer */
 		new_thread->name[CONFIG_THREAD_MAX_NAME_LEN - 1] = '\0';
+		arch_thread_name_set(new_thread, name);
```
Yes, lets do that instead

rugeGerritsen commented on arch/posix/core/thread.c:
```diff
@@ -54,6 +54,17 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+	thread_status = thread->callee_saved.thread_status;
+	thread_index = thread_status->thread_idx;
+
+	return posix_arch_thread_name_set(thread_index, str);
```
Good point. That improves user experience

rugeGerritsen committed:
tests: bluetooth: hci_uart_async: Set thread name after creation

When running on native_posix, the thread needs to be created
before the thread name can be set.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

aescolar reviewed (commented):
Just a name change for the options otherwise this seems good to me.

rugeGerritsen commented on kernel/Kconfig:
```diff
@@ -778,6 +778,11 @@ config ARCH_HAS_CUSTOM_BUSY_WAIT
 	  the provided k_busy_wait(), but instead must do something custom. It must
 	  enable this option in that case.
 
+config ARCH_HAS_CUSTOM_THREAD_NAME
+	bool
+	help
+	  The arch requires an extra function call to apply setting the thread name.
```
I agree, this is a better naming for it

alwa-nordic commented on tests/bluetooth/hci_uart_async/src/test_hci_uart_async.c:
```diff
@@ -70,10 +70,10 @@ static void hci_uart_thread_entry(void *p1, void *p2, void *p3)
 	k_thread_create(&hci_uart_thread, hci_uart_thread_stack,
 			K_THREAD_STACK_SIZEOF(hci_uart_thread_stack), hci_uart_thread_entry, NULL,
 			NULL, NULL, CONFIG_MAIN_THREAD_PRIORITY, 0, K_NO_WAIT);
+	k_thread_name_set(&hci_uart_thread, "hci_uart_main");
```
Not a hard nack, but it used to work to set the name before starting the thread. This has the benefit of the thread always having the correct name from the very start. I suspect it should still work if we don't call `arch_thread_name_set` if the thread is not running yet, and it will be called later from `z_setup_new_thread`.

rugeGerritsen replied:
I agree. This change is not really needed after I've added
```c
	thread_status = thread->callee_saved.thread_status;
	if (!thread_status) {
		return -EAGAIN;
	}
```
Now we always set thread name after creation (checked with spatch)

aescolar replied:
Wasn't this code anyhow incorrect (irrespective of other changes in this PR)?
`k_thread_create()` clears the stored thread name

alwa-nordic replied:
> Wasn't this code anyhow originally incorrect (irrespective of other changes in this PR)? `k_thread_create()` clears the stored thread name

It looks like you are right. Never mind then.

aescolar commented:
> What do we do about the _GNU_SOURCE that the CI does not like?

Nothing. When it is time to merge an admin bypasses the check.

nashif commented:
kernel/arch change, reassigning to kernel maintainers.

Thalley commented on kernel/include/kernel_arch_interface.h:
```diff
@@ -557,6 +557,22 @@ uintptr_t arch_page_info_get(void *addr, uintptr_t *location,
  */
 int arch_printk_char_out(int c);
 
+#ifdef CONFIG_ARCH_HAS_THREAD_NAME_HOOK
+/**
+ * Set thread name hook
+ *
+ * If implemented, any invocation of a function setting a thread name
+ * will invoke this function.
+ *
+ * @param thread    Pointer to thread object
+ * @param str       The thread name
+ *
+ * @retval 0        On success.
+ * @retval -EAGAIN  If the operation could not be performed.
+ */
+int arch_thread_name_set(struct k_thread *thread, const char *str);
+#endif /* CONFIG_ARCH_HAS_THREAD_NAME_HOOK */
```
Any reason to guard the declaration of the function, rather than just the implementation?

rugeGerritsen replied:
I added the ifdefs to the header file to be consistent with other optional APIs provided in this header file.

Thalley commented on kernel/thread.c:
```diff
@@ -145,6 +145,10 @@ int z_impl_k_thread_name_set(k_tid_t thread, const char *str)
 	strncpy(thread->name, str, CONFIG_THREAD_MAX_NAME_LEN - 1);
 	thread->name[CONFIG_THREAD_MAX_NAME_LEN - 1] = '\0';
 
+#ifdef CONFIG_ARCH_HAS_THREAD_NAME_HOOK
+	arch_thread_name_set(thread, str);
+#endif /* CONFIG_ARCH_HAS_THREAD_NAME_HOOK */
```
I believe `IS_ENABLED` can be used here if you don't guard the declaration (which is a good reason not to guard declarations).

rugeGerritsen commented:
Rebased to resolve conflicts

jori-nordic reviewed (approved):
Thank you
![image](https://github.com/user-attachments/assets/202f885d-f213-4d36-bf17-38db7f092b7d)

aescolar commented:
@andyross @peter-mitsis could you take a look?
